include/zephyr/types.h

```diff
@@ -29,6 +29,40 @@ typedef union {
 	void            (*thepfunc)(void);
 } z_max_align_t;
 
+/*
+ * Offset type
+ *
+ * k_off_t, much like the POSIX off_t, is a signed integer type used to represent file sizes and
+ * offsets.
+ *
+ * k_off_t effectively limits the size of files that can be handled by the system, it is
+ * therefore not tied to the word-size of the architecture.
+ *
+ * In order to overcome historical limitations of the off_t type, Zephyr always defines
+ * k_off_t as a 64-bit signed integer.
+ */
+typedef int64_t k_off_t;
```

Comment:
Every `off_t` is changed to `k_off_t`, meaning the system footprint will increase. Is this really correct?

Reply:
Just for illustration, I would normally expect to see `k_off_t` used in thread stacks.

If an application used a new `k_off_t` in every function (which would be unusual), and the stack was 25 function calls deep (which is an exceptionally deep stack), then there would only be 100 bytes of additional stack space used. That average case would be significantly less.

So not too bad.

The benefit is, of course, that Zephyr would always be guaranteed to handle any practically sized file, even if it's 20 GiB or something crazy like that.

Reply:
Moved the Kconfig option to the kernel and renamed `CONFIG_OFFSET_64BIT`, much like `CONFIG_TIMEOUT_64BIT`.

It is not enabled by default, but defers the decision to the user.

At least that sets us up for future-proofing `k_off_t`, @keith-packard.

Reply:
yeah, I get excited when I can reduce a function by four bytes in size... Making the shift from a POSIX type to a Zephyr type is sufficient to resolve the actual bug, we can argue when that type should shift to 64 bits in some later PR. Good call.

Reply:
am I missing where the size is set? This is 64-bit always, which really is not needed for the vast majority of low flash devices zephyr supports

Reply:
Fixed!